### ISSUE TITLE
DNS unable to communicate with API - No route to host

### ISSUE THREAD

**Issue by BenHall** [kubernetes/kubeadm#459]
We're seeing a high number of errors when using Kubeadm (1.7.2) and the DNS being unable to communicate with the API, erroring with `getsockopt: no route to host`

Related to #193

<details>
<summary>Kube Logs</summary>

$ kubectl get svc
NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   10.96.0.1    <none>        443/TCP   2h

$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                    READY     STATUS    RESTARTS   AGE
kube-system   etcd-master                             1/1       Running   1          1h
kube-system   kube-apiserver-master                   1/1       Running   1          1h
kube-system   kube-controller-manager-master          1/1       Running   1          1h
kube-system   kube-dns-2425271678-bq4k8               2/3       Running   81         1h
kube-system   kube-proxy-j7fcx                        1/1       Running   0          1h
kube-system   kube-proxy-njltp                        1/1       Running   1          1h
kube-system   kube-scheduler-master                   1/1       Running   1          1h
kube-system   weave-flux-agent-518873993-4lpsm        1/1       Running   0          6m
kube-system   weave-flux-memcached-2601059440-gt5z2   1/1       Running   0          6m
kube-system   weave-net-85p51                         2/2       Running   4          1h
kube-system   weave-net-tltjm                         2/2       Running   0          1h
sock-shop     carts-3397071306-fkwx5                  1/1       Running   0          1h
sock-shop     carts-db-2910504151-nszdb               1/1       Running   0          1h
sock-shop     catalogue-1527825464-9c7sp              1/1       Running   0          1h
sock-shop     catalogue-db-3344976-xz0dn              1/1       Running   0          1h
sock-shop     front-end-1499397808-srgvq              1/1       Running   0          6m
sock-shop     orders-2016403585-trvs4                 1/1       Running   0          1h
sock-shop     orders-db-4110314047-kchfz              1/1       Running   0          1h
sock-shop     payment-1196669234-xvnsb                1/1       Running   0          1h
sock-shop     queue-master-1626798361-ssdtb           1/1       Running   0          1h
sock-shop     rabbitmq-741378374-pjwkp                1/1       Running   0          1h
sock-shop     shipping-3402079389-v298q               1/1       Running   0          1h
sock-shop     user-4274183869-60qj1                   1/1       Running   0          1h
sock-shop     user-db-3000422539-hln20                1/1       Running   0          1h

$ kubectl logs kube-dns-2425271678-bq4k8 -n kube-system kubedns

E0917 18:42:37.044641       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Service: Get https://10.96.0.1:443/api/v1/services?resourceVersion=0: dialtcp 10.96.0.1:443: getsockopt: no route to host
E0917 18:42:37.044665       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.96.0.1:443: getsockopt: no route to host
I0917 18:42:37.543345       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
I0917 18:42:38.043467       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
I0917 18:42:38.543354       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
I0917 18:42:39.043529       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
I0917 18:42:39.543350       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
I0917 18:42:40.043528       1 dns.go:174] Waiting for services and endpoints to be initialized from apiserver...
E0917 18:42:40.044674       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?resourceVersion=0: dial tcp 10.96.0.1:443: getsockopt: no route to host

</details>

Results of `iptables-save` at https://gist.github.com/BenHall/435bcb11224897b17e9d2f5b596f4a4d

**Comment by BenHall** [comment]
More discussions at https://github.com/heptio/aws-quickstart/issues/69

**Comment by BenHall** [comment]
`curl https://10.96.0.1 -k` works from both the master and node

**Comment by bboreham** [comment]
So that pod is in the host network namespace?
That's not typical for kube-dns

**Comment by BenHall** [comment]
No, the pods are not in a separate namespace. The Pod I got the route from was `weaveworks/weave-npc`

Output from the kube proxy on the host is:
```
default via 172.17.0.1 dev ens3
10.32.0.0/12 dev weave  proto kernel  scope link  src 10.40.0.0
172.17.0.0/16 dev ens3  proto kernel  scope link  src 172.17.0.8
172.18.0.0/24 dev docker0  proto kernel  scope link  src 172.18.0.1
```

This is the Kube DNS Pod:

```
Name:		kube-dns-2425271678-nxpnp
Namespace:	kube-system
Node:		node01/172.17.0.19
Start Time:	Sun, 17 Sep 2017 20:51:08 +0000
Labels:		k8s-app=kube-dns
		pod-template-hash=2425271678
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"kube-system","name":"kube-dns-2425271678","uid":"20dbfdd1-9be9-11e7-9c2d-0242ac11...
		scheduler.alpha.kubernetes.io/critical-pod=
Status:		Running
IP:		10.32.0.17
Created By:	ReplicaSet/kube-dns-2425271678
Controlled By:	ReplicaSet/kube-dns-2425271678
Containers:
  kubedns:
    Container ID:	docker://a4addb34dafb474714a484023491dd1daa0fbf2a342bf558325b02bad7221849
    Image:		gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.4
    Image ID:		docker-pullable://gcr.io/google_containers/k8s-dns-kube-dns-amd64@sha256:40790881bbe9ef4ae4ff7fe8b892498eecb7fe6dcc22661402f271e03f7de344
    Ports:		10053/UDP, 10053/TCP, 10055/TCP
    Args:
      --domain=cluster.local.
      --dns-port=10053
      --config-dir=/kube-dns-config
      --v=2
    State:		Waiting
      Reason:		CrashLoopBackOff
    Last State:		Terminated
      Reason:		Error
      Exit Code:	255
      Started:		Sun, 17 Sep 2017 22:04:06 +0000
      Finished:		Sun, 17 Sep 2017 22:05:06 +0000
    Ready:		False
    Restart Count:	16
    Limits:
      memory:	170Mi
    Requests:
      cpu:	100m
      memory:	70Mi
    Liveness:	http-get http://:10054/healthcheck/kubedns delay=60s timeout=5s period=10s #success=1 #failure=5
    Readiness:	http-get http://:8081/readiness delay=3s timeout=5s period=10s #success=1 #failure=3
    Environment:
      PROMETHEUS_PORT:	10055
    Mounts:
      /kube-dns-config from kube-dns-config (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-j7qtr (ro)
  dnsmasq:
    Container ID:	docker://f776c93682fa24779bb96937fbf47371c72b3cb9935b296a15234bfd423f654f
    Image:		gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.4
    Image ID:		docker-pullable://gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64@sha256:aeeb994acbc505eabc7415187cd9edb38cbb5364dc1c2fc748154576464b3dc2
    Ports:		53/UDP, 53/TCP
    Args:
      -v=2
      -logtostderr
      -configDir=/etc/k8s/dns/dnsmasq-nanny
      -restartDnsmasq=true
      --
      -k
      --cache-size=1000
      --log-facility=-
      --server=/cluster.local/127.0.0.1#10053
      --server=/in-addr.arpa/127.0.0.1#10053
      --server=/ip6.arpa/127.0.0.1#10053
    State:		Running
      Started:		Sun, 17 Sep 2017 22:05:43 +0000
    Last State:		Terminated
      Reason:		Error
      Exit Code:	137
      Started:		Sun, 17 Sep 2017 22:03:37 +0000
      Finished:		Sun, 17 Sep 2017 22:05:43 +0000
    Ready:		True
    Restart Count:	19
    Requests:
      cpu:		150m
      memory:		20Mi
    Liveness:		http-get http://:10054/healthcheck/dnsmasq delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:	<none>
    Mounts:
      /etc/k8s/dns/dnsmasq-nanny from kube-dns-config (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-j7qtr (ro)
  sidecar:
    Container ID:	docker://6c8caea4343800364db338629123e24d642df0d6771a3e3723ef3b0b025cac51
    Image:		gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.4
    Image ID:		docker-pullable://gcr.io/google_containers/k8s-dns-sidecar-amd64@sha256:97074c951046e37d3cbb98b82ae85ed15704a290cce66a8314e7f846404edde9
    Port:		10054/TCP
    Args:
      --v=2
      --logtostderr
      --probe=kubedns,127.0.0.1:10053,kubernetes.default.svc.cluster.local,5,A
      --probe=dnsmasq,127.0.0.1:53,kubernetes.default.svc.cluster.local,5,A
    State:		Waiting
      Reason:		CrashLoopBackOff
    Last State:		Terminated
      Reason:		Error
      Exit Code:	2
      Started:		Sun, 17 Sep 2017 22:05:44 +0000
      Finished:		Sun, 17 Sep 2017 22:07:18 +0000
    Ready:		False
    Restart Count:	21
    Requests:
      cpu:		10m
      memory:		20Mi
    Liveness:		http-get http://:10054/metrics delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:	<none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-j7qtr (ro)
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  kube-dns-config:
    Type:	ConfigMap (a volume populated by a ConfigMap)
    Name:	kube-dns
    Optional:	true
  kube-dns-token-j7qtr:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	kube-dns-token-j7qtr
    Optional:	false
QoS Class:	Burstable
Node-Selectors:	<none>
Tolerations:	CriticalAddonsOnly
		node-role.kubernetes.io/master:NoSchedule
		node.alpha.kubernetes.io/notReady:NoExecute for 300s
		node.alpha.kubernetes.io/unreachable:NoExecute for 300s
Events:
  FirstSeen	LastSeen	Count	From		SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----		-------------			--------	------		-------
  1h		20m		7	kubelet, node01	spec.containers{kubedns}	Warning		Unhealthy	Liveness probe failed: Get http://10.32.0.17:10054/healthcheck/kubedns: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
  1h		8m		73	kubelet, node01	spec.containers{kubedns}	Warning		Unhealthy	Readiness probe failed: Get http://10.32.0.17:8081/readiness: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
  1h		4m		194	kubelet, node01	spec.containers{dnsmasq}	Warning		BackOff		Back-off restarting failed container
  1h		3m		17	kubelet, node01	spec.containers{kubedns}	Normal		Created		Created container
  1h		3m		17	kubelet, node01	spec.containers{kubedns}	Normal		Pulled		Container image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.4" already present on machine
  1h		3m		17	kubelet, node01	spec.containers{kubedns}	Normal		Started		Started container
  1h		2m		30	kubelet, node01	spec.containers{kubedns}	Warning		Unhealthy	Readiness probe failed: Get http://10.32.0.17:8081/readiness: dial tcp 10.32.0.17:8081: getsockopt: network is unreachable
  1h		1m		20	kubelet, node01	spec.containers{dnsmasq}	Normal		Created		Created container
  1h		1m		20	kubelet, node01	spec.containers{dnsmasq}	Normal		Started		Started container
  1h		1m		20	kubelet, node01	spec.containers{dnsmasq}	Normal		Pulled		Container image "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.4" already present on machine
  1h		1m		19	kubelet, node01	spec.containers{dnsmasq}	Normal		Killing		Killing container with id docker://dnsmasq:pod "kube-dns-2425271678-nxpnp_kube-system(ee2b6b4d-9be9-11e7-9c2d-0242ac110008)" container "dnsmasq" is unhealthy, it will be killed and re-created.
  1h		1m		22	kubelet, node01	spec.containers{sidecar}	Normal		Created		Created container
  1h		1m		22	kubelet, node01	spec.containers{sidecar}	Normal		Pulled		Container image "gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.4" already present on machine
  1h		1m		22	kubelet, node01	spec.containers{sidecar}	Normal		Started		Started container
  1h		32s		16	kubelet, node01	spec.containers{dnsmasq}	Warning		Unhealthy	Liveness probe failed: Get http://10.32.0.17:10054/healthcheck/dnsmasq: dial tcp 10.32.0.17:10054: getsockopt: network is unreachable
  1h		28s		68	kubelet, node01	spec.containers{sidecar}	Warning		Unhealthy	Liveness probe failed: Get http://10.32.0.17:10054/metrics: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
  1h		23s		21	kubelet, node01	spec.containers{sidecar}	Warning		Unhealthy	Liveness probe failed: Get http://10.32.0.17:10054/metrics: dial tcp 10.32.0.17:10054: getsockopt: network is unreachable
  1h		23s		22	kubelet, node01	spec.containers{sidecar}	Normal		Killing		Killing container with id docker://sidecar:pod "kube-dns-2425271678-nxpnp_kube-system(ee2b6b4d-9be9-11e7-9c2d-0242ac110008)" container "sidecar" is unhealthy, it will be killed and re-created.
  1h		22s		306	kubelet, node01	spec.containers{kubedns}	Warning		BackOff		Back-off restarting failed container
  1h		22s		365	kubelet, node01					Warning		FailedSync	Error syncing pod
  59m		22s		184	kubelet, node01	spec.containers{sidecar}	Warning		BackOff		Back-off restarting failed container
  1h		18s		65	kubelet, node01	spec.containers{dnsmasq}	Warning		Unhealthy	Liveness probe failed: Get http://10.32.0.17:10054/healthcheck/dnsmasq: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

**Comment by bboreham** [comment]
Looking at your iptables gist, the filer table looks curiously truncated. Not in a way that would obviously break things, but I would reboot the node and let the various components rebuild their rules and then see how it works. 

The logs from the `weave` container, which adds some rules, would be interesting. 

After that, some words on your error message:  "no route to host"; can mean three different things:
1. There is literally no route to that host. Route tables are per-namespace so you need to print the one out from inside kube-dns. 
2. The source and destination are on the same network, but Linux has no arp entry for the destination. Print the arp table inside the namespace. 
3. There is a route, but the source got back an icmp error when it tried to send. This is easy to see with a packet trace. 

Since the destination is a virtual IP, 2 is not supposed to be the case. 
Since your host apparently can make the conversation, we should suspect something between the pod's veth and the host network.

**Comment by BenHall** [comment]
More initial information:
Weave Logs @ https://gist.github.com/BenHall/6496cc1efc6f44bb039b5967b9fc675a

```
$ arp -a (host)
? (172.17.0.1) at 02:42:02:da:e3:1e [ether] on ens3
? (172.17.0.12) at 02:42:ac:11:00:0c [ether] on ens3
? (172.17.0.27) at 02:42:ac:11:00:1b [ether] on ens3

$ arp -a (started by docker start kube dns)
? (10.32.0.1) at <incomplete>  on eth0


after rebooting (different cluster)....

arp -a (dns container)
? (10.32.0.20) at 56:bd:b9:65:40:e4 [ether]  on eth0
? (10.32.0.26) at 36:dd:bd:c0:25:63 [ether]  on eth0
? (10.32.0.29) at de:a3:2e:f2:59:81 [ether]  on eth0
? (10.32.0.1) at 7a:18:cd:ac:41:82 [ether]  on eth0
? (10.32.0.23) at d2:96:37:c5:bb:df [ether]  on eth0
? (10.32.0.24) at 22:9c:f7:00:a4:b6 [ether]  on eth0

arp -a (host)
$ arp -a
? (10.32.0.29) at de:a3:2e:f2:59:81 [ether] on weave
? (172.17.0.1) at 02:42:02:da:e3:1e [ether] on ens3
? (10.32.0.22) at 0e:fa:36:68:a6:12 [ether] on weave
? (10.32.0.25) at 1a:77:21:2b:2e:be [ether] on weave
? (10.32.0.18) at be:fc:95:92:ac:36 [ether] on weave
? (10.32.0.23) at d2:96:37:c5:bb:df [ether] on weave
? (10.32.0.19) at 86:61:a2:33:f3:86 [ether] on weave
? (10.32.0.20) at 56:bd:b9:65:40:e4 [ether] on weave
? (10.32.0.26) at 36:dd:bd:c0:25:63 [ether] on weave
? (10.32.0.21) at 0a:94:be:f2:2e:4a [ether] on weave
? (10.32.0.3) at 16:87:60:22:6a:63 [ether] on weave
? (10.32.0.27) at ce:a1:61:20:d3:7f [ether] on weave
? (10.32.0.17) at 56:53:7e:3e:9e:7b [ether] on weave
? (172.17.0.8) at 02:42:ac:11:00:08 [ether] on ens3
? (10.32.0.24) at 22:9c:f7:00:a4:b6 [ether] on weave
? (172.17.0.26) at 02:42:ac:11:00:1a [ether] on ens3
```

**Comment by bboreham** [comment]
Did you reboot?

**Comment by BenHall** [comment]
Yes, DNS seems to have come up

**Comment by jamiehannaford** [comment]
@bboreham Is there anything which indicates this is a bug with kubeadm? Or is it more of a weave issue?

**Comment by bboreham** [comment]
I disgnosed a conflict between Weave and kube-proxy, provided a fix and am waiting for Ben to say whether it worked.

**Comment by BenHall** [comment]
@bboreham So far so good. It seems to have resolved the issues we were seeing. Thanks so much for the help.

**Comment by bboreham** [comment]
For the record the fix was released in Weave Net 2.0.5

**Comment by jamiehannaford** [comment]
Awesome, thanks @bboreham
